File: dec01/py/part2.py
"""
There are much simpler ways to solve this, but I wanted to experiment with some deques and coroutines in Python.

"""
from collections import deque
depths = deque([int(x.strip()) for x in open('../input.txt', 'r').readlines()])


def process(stream, next_coroutine):
    while(len(stream) >= 3):
        vals = stream[0], stream[1], stream[2]
        next_coroutine.send(vals)
        stream.popleft()
    next_coroutine.close()


def get_depth_increases():
    increases = 0
    previous = None
    try:
        while True:
            current = (yield)
            if previous is None:
                previous = current
                continue # can't compare b/c we don't have a previous value.
            else:
                if sum(current) > sum(previous):
                    increases += 1
            previous = current
    except GeneratorExit:
        print("Done with depth increases!")
        print(f"There were {increases} depth increases")


gdi = get_depth_increases()
gdi.__next__()  # initialize coroutine (run to 'current = (yield)')
process(depths, gdi)

File: dec01/c/part2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


int main()
{
    FILE *fptr;
    int depth_increase_count = 0;
    fptr = fopen("../input.txt", "rb");
    char line [1000];
    int vals[4];
    int count = 0;
    while(fgets(line, sizeof line, fptr) != NULL) {
        count += 1;
        vals[0] = vals[1];
        vals[1] = vals[2];
        vals[2] = vals[3];
        vals[3] = atoi(line);
        if (count < 4) {
            continue;
        }
        if ((vals[0] + vals[1] + vals[2]) < (vals[1] + vals[2] + vals[3])) {
            depth_increase_count++;
        }
    }
    fclose(fptr);
    printf("There were %d depth increases.", depth_increase_count);
    return 0;
};